pkg_install: Support TreeArtifacts.

jacky8hyf opened the pull request:
Implementation notes:
For tree artifacts, when creating directories, we mostly follow the modes set for the whole TreeArtifact, but also +x to allow searching the directory. This is similar to how pkg_tar etc. handles things.

Link: https://github.com/bazelbuild/rules_pkg/issues/308

aiuto commented on pkg/private/install.py.tpl:
```diff
@@ -103,15 +103,27 @@ class NativeInstaller(object):
-        for root, dirs, files in os.walk(entry.src):
-            relative_installdir = os.path.join(entry.dest, root)
+        shutil.copytree(
+            src = entry.src,
```
python style is no space around = for arg passing.

aiuto commented on pkg/private/install.py.tpl:
```diff
@@ -103,15 +103,27 @@ class NativeInstaller(object):
+            dst = entry.dest,
+            symlinks = True,
+            copy_function=lambda s, d: self._do_file_copy(s, d, entry.mode, entry.user, entry.group),
+            dirs_exist_ok = True,
```
I think you should add `ignore_dangling_symlinks=True`.

aiuto commented on pkg/private/install.py.tpl:
```diff
@@ -103,15 +103,27 @@ class NativeInstaller(object):
+        dir_mode = None
+        if entry.mode:
+            dir_mode = int(entry.mode, 8)
+            dir_mode |= 0o111  # +x, allow listing files below directories
```
If we are going to make this choice for them, we might as well make it 555.
Tree artifacts are almost always a bundle of readable things.

aiuto commented on pkg/private/install.py.tpl:
```diff
@@ -103,15 +103,27 @@ class NativeInstaller(object):
+            src = entry.src,
+            dst = entry.dest,
+            symlinks = True,
+            copy_function=lambda s, d: self._do_file_copy(s, d, entry.mode, entry.user, entry.group),
```
Is entry.mode the right thing? That would be the top level dir of the artifact, not a typical file within it.

jacky8hyf replied:
I would assume that `entry` is for the WHOLE tree artifact, so -- for all directories & files within it. I can check what pkg_tar/pkg_zip does and try aligning the behavior too.

jacky8hyf replied:
As far as I can see, the "intermediate directories" within the tree artifact is just set with a hard-coded 755:

https://github.com/bazelbuild/rules_pkg/blob/main/pkg/private/tar/build_tar.py#L285

So I did the same thing here.

For the files within the TreeArtifact, I did use entry.mode. But I haven't double check that's the same behavior in pkg_tar/pkg_zip, so I'll leave this open until I confirm this.

jacky8hyf replied:
I have checked that pkg_tar uses the same behavior here (that is, the files inside the TreeArtifact has the same attribute as the TreeArtifact's `attributes` as well). As such, I think it is doing the right thing now, and I have also added tests to mandate this behavior.

aiuto commented on pkg/private/install.py.tpl:
```diff
@@ -103,15 +103,27 @@ class NativeInstaller(object):
+            dirs_exist_ok = True,
+        )
+
+        dir_mode = None
```
Can you add a comment that this is cleaning up the owners of the directories because copytree will create them as the owner running install.

That *is* what this is doing, right?

jacky8hyf commented:
I did a few changes:

- Clean up some internal functions (done in a separate patch and rebased the patch on top of it)
- Added tests
- Addressed the comments.

As such, I did a force push for a fresh review. Please take a look; thanks!

I didn't handle symlinks within the tree artifact, though.

jacky8hyf commented:
Hello, any updates? Could you please take a look at this pull request? Thanks!

cgrindel reviewed (commented):
LGTM. I will let @aiuto look before merging.
